File: freight-triage-polyglot-ledger-recovery/environment/reconcile/internal/codecx/quoted_freight.go
package codecx

// EncodeQuotedFreight applies the quoted_freight encoding.
func EncodeQuotedFreight(data []byte) []byte {
	out := make([]byte, 0, len(data))
	for _, b := range data {
		if b >= 0x20 && b <= 0x7E && b != '=' {
			out = append(out, b)
			continue
		}
		out = append(out, '=', upperHex[b>>4], upperHex[b&0x0F])
	}
	return out
}

// DecodeQuotedFreight inverts the quoted_freight encoding.
func DecodeQuotedFreight(data []byte) []byte {
	out := make([]byte, 0, len(data))
	for i := 0; i < len(data); i++ {
		if data[i] != '=' {
			out = append(out, data[i])
			continue
		}
		if i+2 >= len(data) {
			break
		}
		high := hexValue(data[i+1])
		low := hexValue(data[i+2])
		if high < 0 || low < 0 {
			return nil
		}
		out = append(out, byte(high<<4|low))
		i += 2
	}
	return out
}

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/codec/QuotedFreight.java
package com.freight.codec;

/** quoted_freight codec. */
public final class QuotedFreight implements Codec {

    private static final String HEX_DIGITS = "0123456789abcdef";
    private static final String UPPER_HEX = "0123456789ABCDEF";
    private static final String BASE32_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
    private static final String BASE64_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

    @Override
    public String name() {
        return "quoted_freight";
    }

    @Override
    public byte[] encode(byte[] data) {
        java.io.ByteArrayOutputStream out = new java.io.ByteArrayOutputStream();
        for (int i = 0; i < data.length; i++) {
            int b = data[i] & 0xFF;
            if (b >= 0x20 && b <= 0x7E && b != '=') {
                out.write(b);
                continue;
            }
            out.write('=');
            out.write(UPPER_HEX.charAt(b >>> 4));
            out.write(UPPER_HEX.charAt(b & 0x0F));
        }
        return out.toByteArray();
    }

    @Override
    public byte[] decode(byte[] data) {
        java.io.ByteArrayOutputStream out = new java.io.ByteArrayOutputStream();
        for (int i = 0; i < data.length; i++) {
            if ((data[i] & 0xFF) != '=') {
                out.write(data[i] & 0xFF);
                continue;
            }
            if (i + 2 >= data.length) {
                break;
            }
            int high = CodecSupport.hexValue((char) (data[i + 1] & 0xFF));
            int low = CodecSupport.hexValue((char) (data[i + 2] & 0xFF));
            if (high < 0 || low < 0) {
                return new byte[0];
            }
            out.write((high << 4) | low);
            i += 2;
        }
        return out.toByteArray();
    }
}
